### PR TITLE
Refactor oauthenticators

### PR DESCRIPTION
**Pull request by GeorgianaElena**
- Creates a base `authenticate` method that consists of:
 ```python
    async def authenticate(self, handler, data=None):
        tokens_info = await self.get_tokens_info(handler)
        user_info = await self.token_to_user(tokens_info)
        auth_state = await self.create_auth_state(tokens_info, user_info)

        return auth_state
```
  The scope is to also introduce `refresh_user` in this workflow
- Implements genric `get_tokens_info` and `token_to_user` methods in `oauth2.py` and lets the oauthenticators only implement `create_auth_state` (this ca be further refactored into smaller pieces)
- Moves some attributes from `generic` to `oauth2`

The puropose of this refactoring is to make implementing https://github.com/jupyterhub/oauthenticator/issues/398 easier.

- [X] oauthenticator.oauth2
- [X] oauthenticator.auth0
- [X] oauthenticator.azuread
- [X] oauthenticator.bitbucket
- [X] oauthenticator.cilogon
- [X] oauthenticator.generic
- [X] oauthenticator.github
- [X] oauthenticator.gitlab
- [x] oauthenticator.globus
- [x] oauthenticator.google
- [x] oauthenticator.okpy
- [x] oauthenticator.openshift
- [x] oauthenticator.mediawiki

### CODE REVIEW

**Comment by minrk** [comment]
Hooray! I've wanted to do this for ages.

**Comment by GeorgianaElena** [comment]
I believe I got myself into a refactoring rabbit hole so I'll stop now and try to provide an update about the state of things so far to help with reviews 😥

### New `authenticate` workflow
I believe I managed to synthesize the `authentication` workflow into a set of methods that can be independently tweaked (or not) by the `oauthenticators` that need some of them done differently. So, here it goes:

1. build the parameters to be used in the request exchanging the oauth code for the access token
2. exchange the oauth code for an access token and get the JSON with info about it
3. use the access_token to get userdata info
4. extract the username out of the user_info dict
5. check if there any refresh_token in the tokens_info dict
6. if not, try to retrieve it from a previous auth_state if available
7. build the auth model to be persisted if authentication goes right
8. check if the username that's authenticating should be authorized
9. update the auth model with any info if available
10. return the auth model

### What changed
(I'm sure I'm going to miss some :/)
- some configs were deprecated in order to be consistent
- new config options were added
- some configs moved from being specific to only one authenticator to being part of the base class
- authenticate now returns more info to satisfy all oauthenticators

### What still needs to be improved

- there's an entire entanglement of methods either raising errors, or returning nulls on failure that should be sorted out
- not enough comments
- not enough corner case checks
- need to sort out making autodoc generate docs for methods too

### What's the connection with `refresh_user`

I believe that `refresh_user` can now be implemented in the base class and some of the newly introduced methods could be reused there.

### For the review process
While I  did mess with the logics of things here and there, the goal of this is to get all oauthenticators to the same common ground without removing or breaking any existing logic.

So because this is a monster PR that touches all authenticators, I would suggest concentrating our efforts on making sure that this doesn't break anything by mistake and we can later further improve the actual logic/code.

The tests pass, but I'm not sure they can be trusted, since they are little :(, hence my fear of breaking anything.

### Apologies
I'm very sorry for the amount of effort to review this PR is generating and I'm happy to provide any extra info, answer any question or change anything to make reviewing easier.

File: oauthenticator/auth0.py [hunk 1]
```diff
@@ -55,14 +58,21 @@ def _auth0_subdomain_default(self):
         return subdomain
 
     username_key = Unicode(
-        os.environ.get("OAUTH2_USERNAME_KEY", "email"),
         config=True,
-        help="Userdata username key from returned json with user data login information",
+        help="Deprecated, use `Auth0OAuthenticator.username_claim`",
     )
```
Comment (consideRatio) [inline oauthenticator/auth0.py:63]: ## Important: breaking change

There is a breaking change related to the `OAUTH2_USERNAME_KEY` environment variable. Adding this line back, setting the default value of this deprecated trait, is not sufficient though.

Our logic associated with the configuration of `_deprecated_oauth_aliases` is working if we via config directly set `username_key`, but not if we set it indirectly via the default value via the `OAUTH2_USERNAME_KEY` environment variable.

Comment (consideRatio) [reply oauthenticator/auth0.py:63]: I think it would be great to drop use of these environment variables and require direct configuration, but preferable not as part of this PR.

Comment (GeorgianaElena) [reply oauthenticator/auth0.py:63]: Yes, I agree! How about we open up an issue about dropping env vars support (if one isnt' open already) after merging this PR and we release them both together as a big breaking version?

Alternatively, I believe we could have the new `username_claim` track OAUTH2_USERNAME_KEY and default to that if not explicitly set as config option?

(Also, I believe there might be another similar situation with another env var)

Comment (consideRatio) [reply oauthenticator/auth0.py:63]: > I believe we could have the new username_claim track OAUTH2_USERNAME_KEY and default to that if not explicitly set as config option?

Let's go for this for now and either an inline FIXME note or dedicated issue on how this is something we should stop supporting this in the future.

**Review by GeorgianaElena — COMMENTED** [review]
Maybe this helps:
- I tried adding some context about the bits that I felt were more sketchy or interesting.  
- Also, one idea for a review approach would be to start with the changes made in the `oauth2.py` file to get a sense of what the general `authenticate` flow is right now, and then check if this looks like it will still  works on all the authenticators. Esp if there are things were left out?

- The way I approached the refactoring was to start from one oauthenticator, then separate the big `authenticate` method into 3 main methods (one that gets the token info, one that get the user info using this token, and one that does all the rest, i.e. returns the expected dict after some processing of the data that needed to go into that), tried to generalize that and moved it to the base class.
  Then, I slowly went through other oauthenticators, and checked if those base methods still made sense for those too. This usually resulted in either adding conditionals, or adding more features/checks to the base methods to make it work for each of the new oauthenticator I was checking
- Unfortunately I heavily relied on tests to make sure the features that were "borrowed" from other oauthenticators (by moving them to the base class) won't break the authentication. But because I don't think the tests cover all edge cases, I cannot be confident the changes here won't break existing configurations under any circumstances

**Comment by yuvipanda** [comment]
We're already running into rebase issues here, unfortunately :(

@consideratio @minrk would either of you have time in say the next two weeks to look at this? If not, I can try find time.

File: oauthenticator/oauth2.py [hunk 2]
```diff
@@ -277,8 +288,45 @@ def _token_url_default(self):
...
+        The field in the userdata response from which to get the JupyterHub username.
+        Examples include: email, username, nickname
+
+        What keys are available will depend on the scopes requested and the authenticator used.
+        """,
+    )
+
+    @default("username_claim")
+    def _username_claim_default(self):
+        return "username"
```
[the remark this answers is not in the document]
Comment (minrk) [reply oauthenticator/oauth2.py:303]: Don't need a dynamic default for a static value. Subclasses can still override with a dynamic default:

```suggestion
    username_claim = Unicode(
        "username",
        config=True,
        help="""Field in userdata reply to use for username
        The field in the userdata response from which to get the JupyterHub username.
        Examples include: email, username, nickname

        What keys are available will depend on the scopes requested and the authenticator used.
        """,
    )

```

File: oauthenticator/oauth2.py [hunk 3]
```diff
@@ -277,8 +288,45 @@ def _token_url_default(self):
...
+
+    # Originally a GenericOAuthenticator only trait
+    userdata_token_method = Unicode(
+        os.environ.get("OAUTH2_USERDATA_REQUEST_TYPE", "header"),
+        config=True,
+        help="Method for sending access token in userdata request. Supported methods: header, url. Default: header",
+    )
+
+    # Originally a GenericOAuthenticator only trait
+    extra_params = Dict(
```
Comment (minrk) [inline oauthenticator/oauth2.py:326]: Should we take this opportunity to call it `token_params` to be consistent with `userdata_params`?

```suggestion
    token_params = Dict(
```

File: oauthenticator/oauth2.py [hunk 4]
```diff
@@ -426,8 +481,274 @@ def get_handlers(self, app):
...
+        Called by the :meth:`oauthenticator.OAuthenticator.token_to_user`
+        """
+        return {
+            "Accept": "application/json",
+            "Content-Type": "application/json",
+            "User-Agent": "JupyterHub",
+            "Authorization": "{} {}".format(token_type, access_token),
+        }
+
+    def build_token_info_req_headers(self):
```
Comment (minrk) [inline oauthenticator/oauth2.py:496]: Consistency with userdata req->request:

```suggestion
    def build_token_info_request_headers(self):
```

File: oauthenticator/oauth2.py [hunk 5]
```diff
@@ -426,8 +481,274 @@ def get_handlers(self, app):
...
+            auth_state = await decrypt(encrypted)
+
+            return auth_state.get('refresh_token')
+        except (ValueError, InvalidToken, EncryptionUnavailable) as e:
+            self.log.warning(
+                f"Failed to retrieve encrypted auth_state for {username}. Error was {e}.",
+            )
+            return
+
+    def build_access_tokens_req_params(self, handler):
```
Comment (minrk) [inline oauthenticator/oauth2.py:559]: Same as above, consistency:

```suggestion
    def build_access_tokens_request_params(self, handler):
```

File: oauthenticator/oauth2.py [hunk 6]
```diff
@@ -426,8 +481,274 @@ def get_handlers(self, app):
...
+            'grant_type': 'authorization_code',
+            'client_id': self.client_id,
+            'client_secret': self.client_secret,
+            'redirect_uri': self.get_callback_url(handler),
+        }
+        params.update(self.extra_params)
+
+        return params
+
+    async def get_tokens_info(self, handler, params):
```
Comment (minrk) [inline oauthenticator/oauth2.py:580]:
```suggestion
    async def get_token_info(self, handler, params):
```

File: oauthenticator/oauth2.py [hunk 7]
```diff
@@ -426,8 +481,274 @@ def get_handlers(self, app):
...
+            the JSON response to the `token_url` the request.
+
+        Called by the :meth:`oauthenticator.OAuthenticator.authenticate`
+        """
+        url = url_concat(self.token_url, params)
+
+        req = HTTPRequest(
+            url,
+            method="POST",
+            headers=self.build_token_info_req_headers(),
```
Comment (minrk) [inline oauthenticator/oauth2.py:594]:
```suggestion
            headers=self.build_token_info_request_headers(),
```

File: oauthenticator/oauth2.py [hunk 8]
```diff
@@ -426,8 +481,274 @@ def get_handlers(self, app):
...
+            body=json.dumps(params),
+            validate_cert=self.validate_server_cert,
+        )
+
+        tokens_info = await self.fetch(req)
```
Comment (minrk) [inline oauthenticator/oauth2.py:599]: Consistency: let's use `token_info` everywhere, instead of plural `tokens_info`.

File: oauthenticator/oauth2.py [hunk 9]
```diff
@@ -426,8 +481,274 @@ def get_handlers(self, app):
...
+
+        if "error_description" in tokens_info:
+            raise web.HTTPError(
+                403,
+                "An access token was not returned: {}".format(
+                    tokens_info["error_description"]
+                ),
+            )
+        elif "access_token" not in tokens_info:
```
Comment (minrk) [inline oauthenticator/oauth2.py:608]: Can this key vary, or is it the same for every provider?

**Review by minrk — COMMENTED** [review]
This is great! A few comments about consistency of things, but the changes all look really great.

File: oauthenticator/oauth2.py (hunk 9 above)
Comment (GeorgianaElena) [reply oauthenticator/oauth2.py]: According to https://www.oauth.com/oauth2-servers/access-tokens/access-token-response/ I believe this is a required key for every oauth2 provider

**Comment by GeorgianaElena** [comment]
Thanks a lot for the feedback. I believe I have addressed all the comments and also rebased the PR.
I have also updated some of the deprecation messages that this introduces to mention `16.0.0` instead of `15.1.0` as the new version when these will be shipped. We can change it later if this version gets released in the meantime though.

**Comment by minrk** [comment]
Terrific job, @GeorgianaElena! I think we should probably do a bit of a write-up in the docs about what changed - in particular noting any config that no longer works or is deprecated due to the normalization of config names.

**Comment by GeorgianaElena** [comment]
Yaaay! Thank you 🚀 I've opened https://github.com/jupyterhub/oauthenticator/issues/549 about the docs part and self-assigned to it.

**Comment by consideRatio** [comment]
Wieee!!!! Amazing effort on this @GeorgianaElena!! And @minrk for going all the way with the associated review efforts!

:Tada: :heart: 🌻

**Comment by consideRatio** [comment]
Following this work, is this known to be breaking or not in some way - besides possible fixes already merged?
